File: test-py/test_discretevariationmodel.py
import pytest
from fontTools.misc.vector import Vector

from fontra.core.classes import DiscreteFontAxis, FontAxis
from fontra.core.discretevariationmodel import DiscreteVariationModel, ErrorDescription

testAxes = [
    FontAxis(
        name="Weight",
        tag="wght",
        label="Weight",
        minValue=400,
        defaultValue=400,
        maxValue=700,
    ),
    DiscreteFontAxis(
        name="Italic", tag="ital", label="Italic", values=[0, 1], defaultValue=0
    ),
]

testLocations = [
    {},
    {"Weight": 700},
    {"Italic": 1},
    {"Weight": 700, "Italic": 1},
]

testSourceData = [
    [0, 0],
    [100, 0],
    [0, 100, 200],  # Incompatible Italic sources
    [100, 100, 300],  # etc.
]


testCases = [
    ({}, [0, 0]),
    ({"Weight": 550}, [50, 0]),
    ({"Weight": 700}, [100, 0]),
    ({"Italic": 1}, [0, 100, 200]),
    ({"Weight": 700, "Italic": 1}, [100, 100, 300]),
    ({"Weight": 550, "Italic": 1}, [50, 100, 250]),
    ({"Weight": 550, "Italic": 0.4}, [50, 0]),
    ({"Weight": 550, "Italic": 0.6}, [50, 100, 250]),
]


@pytest.mark.parametrize("location, expectedResult", testCases)
def test_discreteVariationModel(location, expectedResult):
    assert len(testSourceData) == len(testLocations)
    model = DiscreteVariationModel(testLocations, testAxes)
    deltas = model.getDeltas([Vector(s) for s in testSourceData])
    result = model.interpolateFromDeltas(location, deltas)
    assert result.instance == expectedResult
    assert result.errors is None


testBadLocations = list(testLocations)
testBadLocations[3] = {}

testCasesBadLocations = [
    (
        {},
        [0, 0],
        [
            ErrorDescription(
                message="Italic=0: Locations must be unique.",
                type="model-error",
            ),
        ],
    ),
    (
        {"Weight": 600},
        [100, 0],
        [
            ErrorDescription(
                message="Italic=0: Locations must be unique.",
                type="model-error",
            ),
        ],
    ),
]


@pytest.mark.parametrize(
    "location, expectedResult, expectedErrors", testCasesBadLocations
)
def test_discreteVariationModel_bad_locations(location, expectedResult, expectedErrors):
    model = DiscreteVariationModel(testBadLocations, testAxes)
    deltas = model.getDeltas([Vector(s) for s in testSourceData])
    result = model.interpolateFromDeltas(location, deltas)
    assert result.instance == expectedResult
    assert result.errors == expectedErrors


testIncompatibleSourceData = [
    [0, 0],
    [100, 0, 1, 1],  # bad incompatible source
    [0, 100, 200],
    [100, 100, 300],
]

testCasesIncompatibleSources = [
    (
        {},
        [0, 0],
        [
            ErrorDescription(
                message="",
                type="interpolation-error",
            ),
        ],
    ),
    (
        {"Weight": 600},
        [100, 0, 1, 1],
        [
            ErrorDescription(
                message="",
                type="interpolation-error",
            ),
        ],
    ),
]


@pytest.mark.parametrize(
    "location, expectedResult, expectedErrors", testCasesIncompatibleSources
)
def test_discreteVariationModel_incompatible_sources(
    location, expectedResult, expectedErrors
):
    model = DiscreteVariationModel(testLocations, testAxes)
    deltas = model.getDeltas([Vector(s) for s in testIncompatibleSourceData])
    result = model.interpolateFromDeltas(location, deltas)
    assert result.instance == expectedResult
    assert result.errors == expectedErrors

File: src-js/fontra-core/tests/test-discrete-variation-model.js
import { DiscreteVariationModel } from "@fontra/core/discrete-variation-model.js";
import { expect } from "chai";
import { parametrize } from "./test-support.js";

describe("DiscreteVariationModel tests", () => {
  const testAxes = [
    { name: "Weight", minValue: 400, maxValue: 700, defaultValue: 400 },
    { name: "Italic", values: [0, 1], defaultValue: 0 },
  ];

  const testLocations = [
    {},
    { Weight: 700 },
    { Italic: 1 },
    { Weight: 700, Italic: 1 },
  ];

  const testSourceData = [
    [0, 0],
    [100, 0],
    [0, 100, 200], // Incompatible Italic sources
    [100, 100, 300], // etc.
  ];

  const testCases = [
    {
      location: {},
      expectedResult: [0, 0],
    },
    {
      location: { Weight: 550 },
      expectedResult: [50, 0],
    },
    {
      location: { Weight: 700 },
      expectedResult: [100, 0],
    },
    {
      location: { Italic: 1 },
      expectedResult: [0, 100, 200],
    },
    {
      location: { Weight: 700, Italic: 1 },
      expectedResult: [100, 100, 300],
    },
    {
      location: { Weight: 550, Italic: 1 },
      expectedResult: [50, 100, 250],
    },
    {
      location: { Weight: 550, Italic: 0.4 },
      expectedResult: [50, 0],
    },
    {
      location: { Weight: 550, Italic: 0.6 },
      expectedResult: [50, 100, 250],
    },
  ];

  parametrize("DiscreteVariationModel test", testCases, (testData) => {
    const model = new DiscreteVariationModel(testLocations, testAxes);
    const deltas = model.getDeltas(testSourceData);
    const { instance, errors } = model.interpolateFromDeltas(testData.location, deltas);
    expect(instance).to.deep.equal(testData.expectedResult);
    expect(errors).to.deep.equal(undefined);
  });

  const badLocations = [...testLocations];
  badLocations[3] = {};

  const testCasesBadLocations = [
    {
      location: {},
      expectedResult: [0, 0],
      expectedErrors: [
        {
          message: "Italic=0: locations must be unique",
          type: "model-error",
        },
      ],
    },
    {
      location: { Weight: 600 },
      expectedResult: [100, 0],
      expectedErrors: [
        {
          message: "Italic=0: locations must be unique",
          type: "model-error",
        },
      ],
    },
  ];

  parametrize(
    "DiscreteVariationModel bad locations",
    testCasesBadLocations,
    (testData) => {
      const model = new DiscreteVariationModel(badLocations, testAxes);
      const deltas = model.getDeltas(testSourceData);
      const { instance, errors } = model.interpolateFromDeltas(
        testData.location,
        deltas
      );
      expect(instance).to.deep.equal(testData.expectedResult);
      expect(errors).to.deep.equal(testData.expectedErrors);
    }
  );

  const testIncompatibleSourceData = [
    [0, 0],
    [100, 0, 1, 1], // bad incompatible source
    [0, 100, 200],
    [100, 100, 300],
  ];

  const testCasesIncompatibleSources = [
    {
      location: {},
      expectedResult: [0, 0],
      expectedErrors: [
        {
          message: "arrays have incompatible lengths: 4 != 2",
          type: "interpolation-error",
        },
      ],
    },
    {
      location: { Weight: 600 },
      expectedResult: [100, 0, 1, 1],
      expectedErrors: [
        {
          message: "arrays have incompatible lengths: 4 != 2",
          type: "interpolation-error",
        },
      ],
    },
  ];

  parametrize(
    "DiscreteVariationModel incompatible sources",
    testCasesIncompatibleSources,
    (testData) => {
      const model = new DiscreteVariationModel(testLocations, testAxes);
      const deltas = model.getDeltas(testIncompatibleSourceData);
      const { instance, errors } = model.interpolateFromDeltas(
        testData.location,
        deltas
      );
      expect(instance).to.deep.equal(testData.expectedResult);
      expect(errors).to.deep.equal(testData.expectedErrors);
    }
  );

  const testLocationsNoItalic = [{}, { Weight: 700 }];

  const testSourceDataNoItalic = [
    [0, 0],
    [100, 0],
  ];

  const testCasesNoItalic = [
    {
      location: {},
      expectedResult: [0, 0],
    },
    {
      location: { Weight: 550 },
      expectedResult: [50, 0],
    },
    {
      location: { Weight: 550, Italic: 1 },
      expectedResult: [50, 0],
      expectedErrors: [
        {
          message: "there are no sources for Italic=1",
          type: "model-warning",
        },
      ],
    },
  ];

  parametrize("DiscreteVariationModel no italic", testCasesNoItalic, (testData) => {
    const model = new DiscreteVariationModel(testLocationsNoItalic, testAxes);
    const deltas = model.getDeltas(testSourceDataNoItalic);
    const { instance, errors } = model.interpolateFromDeltas(testData.location, deltas);
    expect(instance).to.deep.equal(testData.expectedResult);
    expect(errors).to.deep.equal(testData.expectedErrors);
  });
});
